pkg/build/controller/buildimagechangetrigger/buildimagechangetrigger_controller.go

```diff
@@ -0,0 +1,293 @@
+		// differs from the last triggered build recorded on the build config for that trigger
+		for _, trigger := range config.Spec.Triggers {
+			if trigger.Type != buildapi.ImageChangeBuildTriggerType {
+				continue
+			}
+			if trigger.ImageChange.From != nil {
+				from = trigger.ImageChange.From
+			} else {
+				from = buildutil.GetInputReference(config.Spec.Strategy)
+			}
+
+			if from == nil || from.Kind != "ImageStreamTag" {
+				continue
+			}
+			fromStreamName, tag, ok := imageapi.SplitImageStreamTag(from.Name)
+			if !ok {
+				glog.Errorf("Invalid image stream tag: %s in build config %s/%s", from.Name, config.Name, config.Namespace)
+				continue
+			}
+
+			fromNamespace := from.Namespace
+			if len(fromNamespace) == 0 {
+				fromNamespace = config.Namespace
+			}
+
+			// only trigger a build if this image stream matches the name and namespace of the stream ref in the build trigger
+			// also do not trigger if the imagerepo does not have a valid DockerImageRepository value for us to pull
+			// the image from
+			if len(stream.Status.DockerImageRepository) == 0 || fromStreamName != stream.Name || fromNamespace != stream.Namespace {
+				continue
+			}
+
+			// This split is safe because ImageStreamTag names always have the form
+			// name:tag.
+			latestReference, ok = imageapi.ResolveLatestTaggedImage(stream, tag)
+			if !ok {
+				glog.V(4).Infof("unable to find tagged image: no image recorded for %s/%s:%s", stream.Namespace, stream.Name, tag)
+				continue
+			}
+			glog.V(4).Infof("Found ImageStream %s/%s with tag %s", stream.Namespace, stream.Name, tag)
+
+			// (must be different) to trigger a build
+			last := trigger.ImageChange.LastTriggeredImageID
+			next := latestReference
+
+			if len(last) == 0 || (len(next) > 0 && next != last) {
+				triggeredImage = next
+				shouldBuild = true
+				// it doesn't really make sense to have multiple image change triggers any more,
+				// so just exit the loop now
+				break
+			}
+		}
+
+		if shouldBuild {
+			glog.V(4).Infof("Running build for BuildConfig %s/%s", config.Namespace, config.Name)
+			buildCauses := []buildapi.BuildTriggerCause{}
+			// instantiate new build
+			request := &buildapi.BuildRequest{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      config.Name,
+					Namespace: config.Namespace,
+				},
+				TriggeredBy: append(buildCauses,
+					buildapi.BuildTriggerCause{
+						Message: buildapi.BuildTriggerCauseImageMsg,
+						ImageChangeBuild: &buildapi.ImageChangeCause{
+							ImageID: latestReference,
+							FromRef: from,
+						},
+					},
+				),
+				TriggeredByImage: &kapi.ObjectReference{
+					Kind: "DockerImage",
+					Name: triggeredImage,
+				},
+				From: from,
+			}
+			// TODO - does bullet 11 of https://github.com/kubernetes/community/blob/master/contributors/devel/controllers.md apply here?
+			// if so, what is the "owner reference" in this context?
```

Comment:
thanks for the confirm / explanation gentlemen - the `TODO` will be removed in the next push